cpp/src/fil/infer.cu

```diff
@@ -373,25 +393,28 @@ struct tree_aggregator_t<NITEMS, CATEGORICAL_LEAF> {
   }
 };
 
-template <int NITEMS, leaf_algo_t leaf_algo, class storage_type>
+template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
```

Comment:
If `cols_in_shmem` is a kernel template parameter, wouldn't it be easier to also add it to `tree_aggregator_t` and remove a lot of changes?

Reply:
I already made the changes, so I assume it's the reviewing effort we're saving?
The end code would be more complicated, because I'd need an extra dispatch level in `get_smem_footprint*` for `cols_in_shmem`.
Also, the convoluted placement code in `tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES>) would not get any simpler. Even if I found a way to split the constructor between specializations, I'd still need to check if the finalize_block fits into `num_cols * NITEMS * sizeof(float)`. This is to avoid a perf regression on the existing (albeit buggy) code.